Contributing to Plone core: skip buildout, use magic (#1936)
plone/documentation: docs/contributing/core/index.md

Reviewer: Line 79 needs a period at the end of its sentence.

```
The current default and development branch of `buildout.coredev` is `6.1`.
```

Also consider either reusing the existing `PLONE_BACKEND_MINOR_VERSION` substitution from `conf.py`, or declaring a new substitution in MyST. I think it's possible, but unlikely, that the two versions would differ. @mauritsvanrees @jensens @davisagli @gforcada might know better than I.

Author: 6.1 -> {PLONE_BACKEND_MINOR_VERSION}

This guide describes the process of how to contribute to, and develop in, Plone core.
It expands upon {doc}`/contributing/index`.

> Although Plone core includes Volto — the React based, default frontend for Plone 6 — this guide does not apply to Volto and its packages.

Reviewer: Please revert. Original punctuation is correct.

To contribute to Volto, see {doc}`../volto`.

This guide assumes that you have basic knowledge of how to use git and GitHub.
If you have never contributed to Plone, or you lack basic knowledge of how to use git and GitHub, you should first read {doc}`/contributing/first-time` for more information.

```{important}
Please {ref}`contributing-sign-and-return-the-plone-contributor-agreement-label` before you contribute.

> This is not optional, but a must before your contribution can be accepted.

Reviewer: Please revert. The original language is deliberate and direct. Less is more.

On Linux, [GNU Compiler Collection (GCC)](https://gcc.gnu.org/) is a common option.

(contributing-core)=
## Install Plone core for development

The tool that installs Plone core is `buildout.coredev`.

> The current default and development branch of `buildout.coredev` is `{PLONE_BACKEND_MINOR_VERSION}`

Reviewer: ```suggestion
The current default and development branch of `buildout.coredev` is `{PLONE_BACKEND_MINOR_VERSION}`.
```

Older versions are named according to their `major.minor` version.
Its versions align with Plone's `major.minor` versions.

Use a separate directory for each version of Plone to which you want to contribute.
This will avoid switching between Git branches, which can cause dependency conflicts between versions of Plone.

To set up a Plone 6 development environment, clone https://github.com/plone/buildout.coredev.

```shell
git clone https://github.com/plone/buildout.coredev
cd buildout.coredev
```

Use the following command to install Plone and run it.

```shell
make run
```

> To create and visit a new Plone site, visit http://localhost:8080, select a distribution, and fill out and submit the form, ensuring you create a site with sample content against which you can test.

Reviewer: Reverting to previous version since this page is intended exclusively for contributing to the Plone core backend.

```suggestion
To create and visit a new Plone site, visit http://localhost:8080, select the Classic UI distribution, and fill out and submit the form, ensuring you create a site with sample content against which you can test.
```

> To create and visit a new Plone site, visit http://localhost:8080, select the Classic UI distribution, and fill out and submit the form, ensuring you create a site with sample content against which you can test.

Reviewer 2: Why Classic UI specifically? I often install the Volto distribution if I'm working on backend changes that I want to test with Volto.

Author: Yes, of course, this should not recommend one of the distributions.

Reviewer: @ksuess @davisagli the instructions on this page do not cover starting the Volto frontend and don't include how to start to develop in Volto. In fact, the second sentence of this page states the following.

> Although Plone core includes Volto—the React based, default frontend for Plone 6—this guide does not apply to Volto and its packages. To contribute to Volto, see [Contributing to Volto](https://plone6--1936.org.readthedocs.build/contributing/volto.html).

Unless and until the instructions on this page actually support Volto development, then we should use the previous version of this sentence, telling developers to select Classic UI.

In my testing, I couldn't even visit the Volto frontend because it doesn't get started. Maybe there is a way to do that, and if there is, then let's get it into this page.

Reviewer 2: @stevepiercy When I'm working on something that involves changes in both Volto and in the backend, I start Volto from the volto repository (as covered in https://plone6--1936.org.readthedocs.build/volto/contributing/developing-core.html#start-plone) and then I start the backend from the buildout.coredev repository (instead of running `make backend-docker-start` from the volto repository, which doesn't give the option to check out backend packages) and then create a site using the Volto distribution.

I'm totally okay with leaving that as out of scope for this PR (let's get this merged), but I think we should explain it somewhere (maybe it belongs under "Contributing to Volto", but it's really about contributing to the backend and then testing those changes with Volto before they're released)

Reviewer 2: (Maybe we should add a `make frontend-docker-start` command to buildout.coredev to parallel `make backend-docker-start` in the Volto repository, in case you just want to get the most recent Volto release running quickly for testing without cloning it separately)

Reviewer: > @stevepiercy When I'm working on something that involves changes in both Volto and in the backend, I start Volto from the volto repository (as covered in https://plone6--1936.org.readthedocs.build/volto/contributing/developing-core.html#start-plone) and then I start the backend from the buildout.coredev repository (instead of running `make backend-docker-start` from the volto repository, which doesn't give the option to check out backend packages) and then create a site using the Volto distribution.

Are you saying that you have two clones, one each for `volto` and `buildout.coredev` as siblings or something?

> I'm totally okay with leaving that as out of scope for this PR (let's get this merged),

Yeah, I think it is out of scope here.

> but I think we should explain it somewhere (maybe it belongs under "Contributing to Volto", but it's really about contributing to the backend and then testing those changes with Volto before they're released)

Are you saying, for example, adding or editing a `plone.restapi` endpoint in your `buildout.coredev` checkout, then running tests in your `volto` checkout? If so, we don't have that in the docs yet. It never occurred to me to have the two checkouts talk to each other. I guess as long as the backend runs on the same host and port that the `volto` checkout requires, it's all good, right?

Reviewer 2: @stevepiercy Yeah, that's right, there's no actual configuration to make the two repositories work with each other, it's just based on the convention of running the backend on port 8080, and Volto expecting one there by default.

On Linux, [GNU Compiler Collection (GCC)](https://gcc.gnu.org/) is a common option.

(contributing-core)=
## Install Plone core for development

The tool that installs Plone core is `buildout.coredev`.

The current default and development branch of `buildout.coredev` is `6.1`

> Former versions are named according to their `major.minor` version.

Reviewer: Please revert. The versions still exist and are active. "Former" implies the version no longer exists or is no longer active. "Older" is a broader term that works. An appropriate usage of "former" would be, "I wish the former President had remained former, instead of getting elected a second time." 😈

Its versions align with Plone's `major.minor` versions.

Use a separate directory for each version of Plone to which you want to contribute.

> This will avoid switching between git branches, which can cause dependency conflicts between versions of Plone.

Reviewer: According to its [README](https://github.com/git/git/blob/master/README.md), "Git" is now capitalized. Originally it was "git", and occasionally "GIT", depending on your mood while using it, as the bulleted list in the README indicates.

```suggestion
This will avoid switching between Git branches, which can cause dependency conflicts between versions of Plone.
```

To set up a Plone 6 development environment, clone https://github.com/plone/buildout.coredev.

```shell
git clone https://github.com/plone/buildout.coredev
cd buildout.coredev
```

Install Plone and run it:

```shell
make run
```

> See more `make` options/targets in the READMEs of the `buildout.coredev` repo.

Reviewer: This file?

https://github.com/plone/buildout.coredev/blob/6.1/README-make.md

Its versions align with Plone's `major.minor` versions.

Use a separate directory for each version of Plone to which you want to contribute.
This will avoid switching between Git branches, which can cause dependency conflicts between versions of Plone.

To set up a Plone 6 development environment, clone https://github.com/plone/buildout.coredev.

```shell
git clone https://github.com/plone/buildout.coredev
cd buildout.coredev
```

> Install Plone and run it:

Reviewer: ```suggestion
Use the following command to install Plone and run it.
```

```shell
make run
```

> 

Reviewer: ```suggestion

To create and visit a new Plone site, visit http://localhost:8080, select the Classic UI distribution, and fill out and submit the form, ensuring you create a site with sample content against which you can test.

Stop Plone with the keyboard shortcut {kbd}`ctrl-c`.

```

> See more `make` options/targets in the READMEs of the `buildout.coredev` repo: https://github.com/plone/buildout.coredev/blob/6.1/README-make.md

Reviewer: "Make" is the brand name, and it is not an inline literal.

````suggestion
```{seealso}
For more Make targets, see documentation in [`buildout.coredev`'s `README-make.md` file](https://github.com/plone/buildout.coredev/blob/6.1/README-make.md).
The most relevant section is [How to use the `Makefile`](https://github.com/plone/buildout.coredev/blob/6.1/README-make.md#how-to-use-the-makefile).
```
````

Only a few packages are in {file}`src/` by default.

Next edit file {file}`mxcheckouts.ini`, and add the name of the package that you want to develop.

> For example `plone.app.multilingual`:

Reviewer: ```suggestion
Next, edit the file {file}`mxcheckouts.ini` at the root of your checkout, adding the name of the package that you want to develop as a new section.
The following example adds `plone.app.multilingual`.
```

```ini
[plone.app.multilingual]
use = true
```

> You can now check out the package and start your instance with the checked out package:

Reviewer: ```suggestion
You can now check out the package and start your instance with the checked out package using the following commands.
```

## Test locally

> If you change the expected behavior of a feature in a package, you should write a unit and/or acceptance test to cover the change.

Reviewer: ```suggestion
If you change the expected behavior of a feature in a package, you should write a unit or acceptance test, as appropriate, to cover the change.
```

Plone uses [Playwright](https://playwright.dev/) to run acceptance tests.
`plone.app.robotframework` provides a script to install Playwright browsers.

```shell
.venv/bin/rfbrowser init
```

To run tests for the specific package that you modified, use `.venv/bin/pytest` followed by the package name, as shown in the following example.

```shell
.venv/bin/pytest src/plone.app.multilingual
```

If any test fails, do not commit and push the changes.
Instead write a test that passes.

> You can run all tests for the package you are developing or just unit tests or just acceptance tests.

Reviewer: ```suggestion
Following [How to invoke pytest](https://docs.pytest.org/en/stable/how-to/usage.html), you can run all tests for the package you're developing, only unit tests, or only acceptance tests, using the pytest keyword option `-k` followed by a string expression in double quotes, as shown below.
```

```shell
.venv/bin/pytest src/plone.app.multilingual -v -k "not robot" 
```

```shell
.venv/bin/pytest src/plone.app.multilingual -v -k "robot" 
```

> After the package level tests pass with your change, you can {ref}`contributing-core-create-a-pull-request-label` and let CI run and ask Jenkins to run the full test suite.

Reviewer: ```suggestion
After the package level tests pass with your change, you can {ref}`contributing-core-create-a-pull-request-label`, which will start the CI checks.
After all CI checks pass, then you can ask Jenkins to run the full test suite.
```

> However, if CI or Jenkins report a test failure that you want to troubleshoot locally, you can run the full test suite to ensure other packages aren't affected by the change.

Reviewer: ```suggestion
If either CI or Jenkins report a test failure, and you want to troubleshoot it locally, then you can run the full test suite to ensure other packages aren't affected by the change.
```

> It takes a while to run the full unit test suite.

Reviewer: ```suggestion
It takes about 15-20 minutes to run the full unit test suite.
```

```shell

> .venv/bin/pytest

Reviewer: In my checkout, I see `src/docs`, which is the `plone/documentation` repo. It looks like it got checked out as well. This causes the above command to emit errors, such as the following.

```console
____________________________________________________________________________________________________________________________ ERROR collecting src/docs/submodules/plone.api/src/plone/api/tests/test_addon.py ____________________________________________________________________________________________________________________________
import file mismatch:
imported module 'plone.api.tests.test_addon' has this __file__ attribute:
  /PATH-TO/buildout.coredev/.venv/lib/python3.13/site-packages/plone/api/tests/test_addon.py
which is not the same as the test file we want to collect:
  /PATH-TO/buildout.coredev/src/docs/submodules/plone.api/src/plone/api/tests/test_addon.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

I used the following command to work around it.

```
$ .venv/bin/pytest --ignore=src/docs 
```

I don't know if this a good idea, or whether the documentation repo should be checked out in the first place. Maybe @jensens can advise?

## Test locally

If you change the expected behavior of a feature in a package, you should write a unit or acceptance test, as appropriate, to cover the change.

Plone uses [Playwright](https://playwright.dev/) to run acceptance tests.
`plone.app.robotframework` provides a script to install Playwright browsers.

```shell
.venv/bin/rfbrowser init
```

To run tests for the specific package that you modified, use `.venv/bin/pytest` followed by the package name, as shown in the following example.

```shell
.venv/bin/pytest src/plone.app.multilingual
```

If any test fails, do not commit and push the changes.
Instead write a test that passes.

> Following [How to invoke pytest](https://docs.pytest.org/en/stable/how-to/usage.html), you can run all tests for the package you're developing, only unit tests, or only acceptance tests, using the pytest keyword option `-k` followed by a string expression in double quotes, as shown below.

Reviewer: Meh, I missed the mark on my first attempt. I had to rewrite this to align with the order of commands. I want to preserve the order because I think it is more helpful to inform users to focus first on the things they change than the entire test suite. Also this version has shorter, clearer sentences.

```suggestion
Following [How to invoke pytest](https://docs.pytest.org/en/stable/how-to/usage.html), you can run tests for the package you're developing.
Using the pytest keyword option `-k`, followed by a string expression in double quotes, you can select only unit tests or only acceptance tests to run as shown below.
With no options, pytest will run all tests.
```
